private/pkg/storage/filter.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"io/fs"
+
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+)
+
+// FilterReadBucket filters the ReadBucket.
+//
+// If the Matchers are empty, the original ReadBucket is returned.
+// If there is more than one Matcher, the Matchers are anded together.
+func FilterReadBucket(readBucket ReadBucket, matchers ...Matcher) ReadBucket {
```

Comment:
Chose `FilterReadBucket` over `MatchReadBucket`, the naming of filter with the filter funcs being called matchers I think is still clear.